walk/cache.go

```diff
@@ -0,0 +1,73 @@
+package walk
+
+import (
+	"fmt"
+	"sync"
+)
+
+// cache is an in-memory cache for file system information. Its purpose is to
+// speed up walking over large directory trees (commonly, the entire repo)
+// by parallelizing parts of the walk while still allowing random access
+// to parts of the directory tree that haven't been loaded yet.
+type cache struct {
+	entryMap sync.Map
+}
+
+type cacheEntry struct {
```

Comment:
Is memory usage of the cache a potential concern? Depending on the padding of Go structs, we could reduce the footprint by storing only one of info and err (via two different entry types).